File: erlang/src/common_lib.erl
-module(common_lib).
-export([copy_file/1, copy_stream/2]).

copy_stream(In, Out) ->
    case file:read(In, 65536) of
        eof ->
            ok;
        {ok, Bin} ->
            ok = file:write(Out, Bin),
            copy_stream(In, Out);
        {error, Reason} ->
            {error, Reason}
    end.

copy_file(Path) ->
    case file:open(Path, [read, binary]) of
        {ok, In} ->
            Result = copy_stream(In, standard_io),
            ok = file:close(In),
            Result;
        {error, Reason} ->
            {error, Reason}
    end.

File: python/src/common_lib.py
#!/usr/bin/env python3
# Copyright (C) 2026 Lenik <zephyr@bodz.net>
# SPDX-License-Identifier: AGPL-3.0-or-later

from __future__ import annotations

import gettext
import locale
import os
from pathlib import Path
from typing import BinaryIO

TEXT_DOMAIN = "puff1"


def init_i18n(argv0: str) -> gettext.NullTranslations:
    locale.setlocale(locale.LC_ALL, "")

    localedir = os.environ.get("ZEPHYR_LOCALEDIR")
    if not localedir and "/" in argv0:
        build_po = Path(argv0).resolve().parent / "po"
        if build_po.is_dir():
            localedir = str(build_po)

    trans = gettext.translation(TEXT_DOMAIN, localedir=localedir, fallback=True)
    trans.install()
    return trans


def copy_stream(src: BinaryIO, dst: BinaryIO) -> None:
    while True:
        chunk = src.read(8192)
        if not chunk:
            return
        dst.write(chunk)
        dst.flush()


def copy_file(path: str, out: BinaryIO) -> None:
    with open(path, "rb") as fh:
        copy_stream(fh, out)
